crates/turbopack-core/src/source_map.rs

```diff
@@ -0,0 +1,303 @@
+                    // The sourcemap package incorrectly returns the last token for large lookup
+                    // lines.
+                    Some(t) if t.get_dst_line() == line as u32 => Some::<Token>(t.into()),
+                    _ => None,
+                }
+            }
+
+            SourceMap::Sectioned(map) => {
+                let len = map.sections.len();
+                let mut low = 0;
+                let mut high = len;
+                let pos = SourcePos { line, column };
+
+                // A "greatest lower bound" binary search. We're looking for the closest section
+                // offset <= to our line/col.
+                while low < high {
+                    let mid = (low + high) / 2;
+                    if pos < map.sections[mid].offset {
+                        high = mid;
+                    } else {
+                        low = mid + 1;
+                    }
+                }
+
+                // Our GLB search will return the section immediately to the right of the
+                // section we actually want to recurse into, because the binary search does not
+                // early exit on an exact match (it'll `low = mid + 1`).
+                if low > 0 && low <= len {
+                    let SourceMapSection { map, offset } = &map.sections[low - 1];
+                    // We're looking for the position `l` lines into region covered by this
+                    // sourcemap s section.
+                    let l = line - offset.line;
+                    // The source map starts if offset by the column only on its first line. On
+                    // the 2nd+ line, the section covers starting at column 0.
+                    let c = if line == offset.line {
+                        column - offset.column
+                    } else {
+                        column
+                    };
+                    return Ok(map.lookup_token(l, c));
+                }
+                None
+            }
+        };
+        Ok(OptionToken(token).cell())
+    }
+}
+
+/// A regular source map covers an entire file.
+#[derive(Debug, Serialize, Deserialize)]
+pub struct RegularSourceMap(Arc<CrateMapWrapper>);
+
+impl RegularSourceMap {
+    fn new(map: CrateMap) -> Self {
+        RegularSourceMap(Arc::new(CrateMapWrapper(map)))
+    }
+}
+
+impl Deref for RegularSourceMap {
+    type Target = Arc<CrateMapWrapper>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl Eq for RegularSourceMap {}
+impl PartialEq for RegularSourceMap {
+    fn eq(&self, other: &Self) -> bool {
+        Arc::ptr_eq(&self.0, &other.0)
+    }
+}
+
+/// Wraps the CrateMap struct so that it can be cached in a Vc.
+///
+/// CrateMap contains a raw pointer, which isn't Send, which is required to
+/// cache in a Vc. So, we have wrap it in 4 layers of cruft to do it. We don't
+/// actually use the pointer, because we don't perform sourcesContent lookups,
+/// so it's fine.
```

Comment:
This disclaimer could go in a `// Safety` comment above the below `unsafe impl`s.

Reply:
Done.